[ACIX-1017] Migrate Omnibus guide to the docs site (#41013)
DataDog/datadog-agent: docs/public/how-to/build/distributions.md

The path you pass with the `--base-dir` option will be used as a working directory for the Omnibus build. Once the build completes, it will contain:

| Directory | Contents                                         |
| --------- | ------------------------------------------------ |
| `src`     | The sources downloaded by Omnibus                |
| `cache`   | The binaries cached after building those sources |
| `pkg`     | The final `deb`/`rpm`/`pkg` artifacts            |

??? note "Make sure to pass a `--base-dir` !"
    It is strongly advised to pass a `--base-dir`, and point it to a directory **outside of the Agent repo.**

    By default Omnibus stores packages in the project folder itself: running the task multiple times would recursively add those artifacts to the source files for the `datadog-agent` software definition.

/// tip
You can fine tune an Omnibus run by passing more options, see `dda inv -- omnibus.build --help` for the list of all the available options.

You can chose to generate an installable package in the form of a `deb`/`rpm` artifact by providing a `OMNIBUS_FORCE_PACKAGES` environment variable during the build.

> > On macOS, a `dmg` artifact will always be generated.

Author: Is it a `dmg` or a `pkg` ? The docs are inconsistent on this

Reviewer: It's a `dmg` they are stored [here](https://s3.amazonaws.com/dd-agent).
Should it be on line 170?

# How to build agent distribution packages

---
## Intro
### What are distribution packages ?

By "distribution package", we mean an artifact used by end users to install the Agent on their system. The format is OS-specific:

| OS      | Package Format                              |
| ------- | ------------------------------------------- |
| Linux   | `.deb` (Debian-based) / `.rpm` (RHEL-based) |
| Windows | `.msi`                                      |
| macOS   | `.pkg`                                      |

These distribution packages contain [a binary of the Agent](standalone.md), along with any supporting libraries needed for the agent to function properly.

### Omnibus

Agent packages for all the supported platforms are built using [Omnibus](https://github.com/chef/omnibus).

> /// info

Reviewer: Is this a valid markdown indication?

Reviewer: wow I looked at the wrong place, thanks!

There is an ongoing effort to migrate our build system to [Bazel](https://bazel.build/), so this may change in the near-to-mid future.
///

/// warning
Omnibus creates a package **for the operating system it runs on**, so you'll get a `.deb` package on Debian-based distros, an `msi` installer on Windows etc.

There is currently no way to "cross-build" packages for a platform different than the host's.
///

Omnibus is best called indirectly, via [dda](../../../setup/required/#tooling) commands.
The main entrypoint is the `omnibus.build` invoke task, which you can run like this:
```bash
dda inv omnibus.build
```
> This will probably not work out of the box though - see instructions below for more information.

## Building for Linux: `deb` and `rpm`
### Containerized build (recommended)

> We provide a Docker image containing all the build dependencies required for building Omnibus Linux packages: [`datadog/agent-buildimages-linux`](https://hub.docker.com/r/datadog/agent-buildimages-linux)

Reviewer: I *think* we still cannot build the rpm packages on these images?

Author: Right, forgot about that. I'll add a section about the `rpm` image with a note saying it will probably be merged in the dev env soon

> ??? tip "Building the image locally"

Reviewer: Is this format a valid markdown one?

Author: It's not markdown itself, but it is valid for the material mkdocs framework we use (specifically it's a [collapsible admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#collapsible-blocks))

    The Dockerfile for this image is available in the [datadog-agent-buildimages](https://github.com/DataDog/datadog-agent-buildimages) repository.
    To build it from scratch, you can run the following command from the root of that repo:

    ```bash
    docker build -t datadog-agent-buildimages:linux -f linux/Dockerfile .
    ```

1. Make sure Docker is running on your machine.

> 1. Navigate to the root folder of a clone of the `datadog-agent` repo.
> 1. Run the following command, which will create a container for the previously-mentioned image, and run the `omnibus.build` task inside.

Reviewer: ```suggestion
2. Navigate to the root folder of a clone of the `datadog-agent` repo.
3. Run the following command, which will create a container for the previously-mentioned image, and run the `omnibus.build` task inside.
```

```bash
docker run
    -v "$PWD:/go/src/github.com/DataDog/datadog-agent"
    -v "/tmp/omnibus:/omnibus"
    -v "/tmp/opt/datadog-agent:/opt/datadog-agent"
    -v"/tmp/gems:/gems"
    --workdir=/go/src/github.com/DataDog/datadog-agent

>     datadog/agent-buildimages-linux-glibc-2-17-x64

Reviewer: ```suggestion
    datadog/agent-buildimages-linux
```

Running Omnibus builds locally may affect the global state of your machine, and in particular the installation of the Agent already present on your laptop.

Please use the [containerized build](#containerized-build-recommended) instead.
///

Running an Omnibus build will both create and install an Agent distribution package.

* The project will be built locally into a `.tar.xz` archive under `omnibus/pkg`.
* The project will be installed under `/opt/datadog-agent`. This is the same path where the Agent is installed on customer machines.

/// warning
If you already have a Datadog Agent installed, you will need to move it to a different location before operating Omnibus - otherwise it will get overwritten by the build.
> As a Datadog employee, an Agent is installed on your machine during IT's onboarding session.
///

??? note "Linux-specific requirements"
    * On Linux, you will need root privileges, as you need permission to write into `/opt`
    * On Linux, some configuration files will also be dropped under `/etc`.

1. Follow the [general local setup instructions](../../../setup/manual)

> 1. Make `/opt` world-readable
> 1. Run the following command:

Reviewer: ```suggestion
2. Make `/opt` world-readable
3. Run the following command:
```

Author: See [above](https://github.com/DataDog/datadog-agent/pull/41013#discussion_r2358385529)

```
dda inv -- omnibus.build --base-dir=$HOME/.omnibus
```

> > On MacOS, you might want to skip the signing step by adding the `--skip-sign` flag.

Reviewer: I suppose this part should be on the paragraph you start at line 134?

Author: Not a bad idea, I'll move the macos-related tips down there :+1:

| Windows | `.msi`                                      |
| macOS   | `.pkg`                                      |

These distribution packages contain [a binary of the Agent](standalone.md), along with any supporting libraries needed for the agent to function properly.

### Omnibus

Agent packages for all the supported platforms are built using [Omnibus](https://github.com/chef/omnibus).
/// info
There is an ongoing effort to migrate our build system to [Bazel](https://bazel.build/), so this may change in the near-to-mid future.
///

/// warning
Omnibus creates a package **for the operating system it runs on**, so you'll get a `.deb` package on Debian-based distros, an `msi` installer on Windows etc.

There is currently no way to "cross-build" packages for a platform different than the host's.
///

Omnibus is best called indirectly, via [dda](../../../setup/required/#tooling) commands.
The main entrypoint is the `omnibus.build` invoke task, which you can run like this:
```bash
dda inv omnibus.build
```
> This will probably not work out of the box though - see instructions below for more information.

## Building for Linux: `deb` and `rpm`
### Containerized build (recommended)

> We provide a Docker image containing all the build dependencies required for building Omnibus Linux packages: [`datadog/agent-buildimages-linux`](https://hub.docker.com/r/datadog/agent-buildimages-linux)

Reviewer: I think so as it provides additional amenities. We should redirect to the user guide of the devenv images as well if we want them to benefit from it.

    ```bash
    docker build -t datadog-agent-buildimages:linux -f linux/Dockerfile .
    ```

1. Make sure Docker is running on your machine.
2. Navigate to the root folder of a clone of the `datadog-agent` repo.
3. Run the following command, which will create a container for the previously-mentioned image, and run the `omnibus.build` task inside.
```bash
docker run
    -v "$PWD:/go/src/github.com/DataDog/datadog-agent"
    -v "/tmp/omnibus:/omnibus"
    -v "/tmp/opt/datadog-agent:/opt/datadog-agent"
    -v"/tmp/gems:/gems"
    --workdir=/go/src/github.com/DataDog/datadog-agent
    datadog/agent-buildimages-linux-glibc-2-17-x64
    dda inv -- -e omnibus.build --base-dir=/omnibus --gem-path=/gems
```

/// info
The container will bind-mount 3 volumes on the host to avoid starting from scratch
at each Omnibus run:

 * `/tmp/omnibus`, containing the Omnibus base directory
 * `/tmp/opt/datadog-agent`, containing the Omnibus installation directory
 * `/tmp/gems`, containing all the ruby gems installed with Bundler
///

///bug
If the build image crashes when running it on a modern Linux distribution, you might be
affected by [this bug](https://github.com/moby/moby/issues/28705).

> ///

Reviewer: Yes please remove this if it's invalid, thanks
